sound/soc/sof/intel/hda-ipc.c

```diff
@@ -254,12 +254,13 @@ int hda_dsp_ipc_get_window_offset(struct snd_sof_dev *sdev, u32 id)
 }
 
 void hda_ipc_msg_data(struct snd_sof_dev *sdev,
-		      struct snd_pcm_substream *substream,
+		      struct snd_sof_pcm_stream *sps,
 		      void *p, size_t sz)
 {
-	if (!substream || !sdev->stream_box.size) {
+	if (!sps || !sdev->stream_box.size) {
 		sof_mailbox_read(sdev, sdev->dsp_box.offset, p, sz);
 	} else {
+		struct snd_pcm_substream *substream = sps->substream;
```

Comment:
is it possible that sps->substream is NULL?
we had that check prior this change.

Reply:
@ujfalusi from my investigation sps->substream cannot be NULL. 

Firstly, it was previously tested for NULL because it was used as a convention to know from which mailbox to read.

```
ipc.c:369:      snd_sof_ipc_msg_data(sdev, NULL, &hdr, sizeof(hdr));
ipc.c:425:      snd_sof_ipc_msg_data(sdev, NULL, &posn, sizeof(posn));
ipc.c:456:      snd_sof_ipc_msg_data(sdev, stream, &posn, sizeof(posn));
ipc.c:487:      snd_sof_ipc_msg_data(sdev, stream, &posn, sizeof(posn));
```

Secondly, when a PCM is allocated via:
```
int snd_pcm_new(struct snd_card *card, const char *id, int device, 
int playback_count, int capture_count, struct snd_pcm **rp)
```
Each substream is allocated like this:
```
sound/core/pcm.c: snd_pcm_new_stream: 665
  for (idx = 0, prev = NULL; idx < substream_count; idx++) {
»       »       substream = kzalloc(sizeof(*substream), GFP_KERNEL);

```

Thirdly, if we follow the code around:

```
ipc.c:456:      snd_sof_ipc_msg_data(sdev, stream, &posn, sizeof(posn));
ipc.c:487:      snd_sof_ipc_msg_data(sdev, stream, &posn, sizeof(posn));
```

We will notice that `stream->substream` will be accessed without any NULL check.

E.g

```
ipc.c:456:      snd_sof_ipc_msg_data(sdev, stream, &posn, sizeof(posn));
 -> snd_sof_pcm_period_elapsed(stream->substream);
      -> asoc_substream_to_rtd(substream);
          -> #define snd_pcm_substream_chip(substream) ((substream)->private_data)                                                                                                                                      
```